app/jobs/order_cycle_notification_job.rb

```diff
@@ -7,5 +7,6 @@ def perform(order_cycle_id)
     order_cycle.suppliers.each do |supplier|
       ProducerMailer.order_cycle_report(supplier, order_cycle).deliver_now
     end
+    order_cycle.update_columns mails_sent: true
```

Comment:
Double sending could be a problem if this process takes longer and we run into a race condition. We could load the OC with a pessimistic lock above but a lengthy process like sending mail shouldn't block anything else...

Let's leave it for now but maybe we need another mechanism that prevents several jobs of the same type to run at the same time. Is there a Sidekiq option for that?

Reply:
In this case it's the `OrderCycleClosingJob` which actually controls whether or not to send the notifications. The `mails_sent` flag is only really used in the UI and not in the email sending logic.

`OrderCycleClosingJob` figures out which recently closed OCs need to have notifications sent and enqueues the `OrderCycleNotificationJob` for each of them using `perform_later` (which is non-blocking) and then immediately marks all the relevant OCs with a `processed_at` timestamp. 

Given that flow, I don't think we'll get any double-sending as `OrderCycleClosingJob` runs once every 5 minutes and I don't think it can take more than a second to complete. :thinking: